### PR TITLE
Store in localstorage viewed products and display unread badge

### LINKED ISSUE #11: Display unread products badge
Store seen products somewhere in cookies / local storage to display an unread badge on product that the user havn't seen yet

### PR DESCRIPTION
**Pull request by ErenMustafaGit**
Closes #11

### CODE REVIEW

**Comment by ErenMustafaGit** [comment]
Add the explanation of this update in app notification (strapi side) @FranckG28

File: components/home/card.tsx [hunk 1]
```diff
@@ -1,34 +1,67 @@
...
 }) {
+  const cardRef: any = useRef();
+  const entry = useIntersectionObserver(cardRef, {
+    freezeOnceVisible: true,
+    threshold: 0.8,
+  });
+  const isVisible = !!entry?.isIntersecting;
+  useEffect(() => {
+    if (isVisible) {
+      const value = window.localStorage.getItem("vine-viewed-products");
```
Comment (FranckG28) [inline components/home/card.tsx:39]: maybe use a hooks to manage localStorage.
[useLocalStorage](https://usehooks-ts.com/react-hook/use-local-storage) is already included in useHooks-ts ;)

To go further if you dare, you can even wrap it in a useViewedProducts hooks that provide an interface like this for example

```ts
const { viewedProducts, addViewedProducts } = useViewedProducts

//[...]
addViewedProducts(id);

```

File: package.json [hunk 2]
```diff
@@ -35,7 +35,8 @@
     "react-markdown": "^8.0.4",
     "react-wrap-balancer": "^0.3.0",
     "typescript": "4.9.4",
-    "use-debounce": "^9.0.3"
+    "use-debounce": "^9.0.3",
```
Comment (FranckG28) [inline package.json:38]: Can we remove it and use the one from usehooks-ts instead ?

Comment (ErenMustafaGit) [reply package.json:38]: I don't know, it was already used. The line changed only with the comma.
Used in search-input.tsx.
I think it is a good idea to use the usehooks-ts version (but not priority)

File: components/home/card.tsx (hunk 1 above)
Comment (ErenMustafaGit) [reply components/home/card.tsx:39]: Hand-made hooks and useHooks-ts hooks for localstorage are not working. Setting the localstorage simultanly and using hooks are not working, it only add 1 item from 3 hooks call...
image

Comment (ErenMustafaGit) [reply components/home/card.tsx:39]: I did this code because i already tried to do a custom hook but it didn't work the first time

Comment (ErenMustafaGit) [reply components/home/card.tsx:39]: Different version of hook : 

Without external hook : 
```ts
export default function useViewedProducts() {
  const viewedProducts = JSON.parse(
    window.localStorage.getItem("vine-viewed-products") || "[]",
  );

  const addViewedProduct = (productId: never) => {
    if (viewedProducts.includes(productId)) return;

    const newViewedProducts = [productId, ...viewedProducts];
    window.localStorage.setItem(
      "vine-viewed-products",
      JSON.stringify(newViewedProducts),
    );
  };

  return {
    viewedProducts,
    addViewedProduct,
  };
}
```

With useState hook :
```ts
import { useEffect, useState } from "react";

export default function useViewedProducts() {
  const [viewedProducts, setViewedProducts] = useState([]);

  useEffect(() => {
    if (typeof window !== "undefined") {
      const viewedProducts = JSON.parse(
        window.localStorage.getItem("vine-viewed-products") || "[]",
      );
      setViewedProducts(viewedProducts);
    } else {
      setViewedProducts([]);
    }
  }, []); // Empty array ensures that effect is only run on mount*/

  const addViewedProduct = (productId: never) => {
    if (viewedProducts.includes(productId)) return;
    const newViewedProducts = [productId, ...viewedProducts];
    setViewedProducts(newViewedProducts);
  };

  return {
    viewedProducts,
    setViewedProducts,
    addViewedProduct,
  };
}
```

With useHooks-ts
```ts
import { useLocalStorage } from "usehooks-ts";

export default function useViewedProducts() {
  const [viewedProducts, setViewedProducts] = useLocalStorage(
    "vine-viewed-products",
    [],
  );

  const addViewedProduct = (productId: never) => {
    if (viewedProducts.includes(productId)) return;
    const newViewedProducts = [productId, ...viewedProducts];
    setViewedProducts(newViewedProducts);
  };

  return {
    viewedProducts,
    setViewedProducts,
    addViewedProduct,
  };
}
```

Comment (FranckG28) [reply components/home/card.tsx:39]: Ohh ! The last one "With useHooks-ts" is my prefered one, but does it works ? It's weird that a hooks by a popular library like this one doesn't work 🤔

File: package.json (hunk 2 above)
Comment (FranckG28) [reply package.json]: I've made it ;)

File: components/home/card.tsx (hunk 1 above)
Comment (ErenMustafaGit) [reply components/home/card.tsx:39]: It does not work when we use this hook at the same time in different component

File: components/home/card.tsx [hunk 3]
```diff
@@ -1,34 +1,67 @@
...
   className,
 }: {
+  id: number;
   title: string;
   description: string;
   demo: ReactNode;
   large?: boolean;
   className?: string;
 }) {
+  const cardRef: any = useRef();
```
[the remark this answers is not in the document]
Comment (ErenMustafaGit) [reply components/home/card.tsx]: Done in wrapper

File: components/home/intersection-observer-wrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useRef, cloneElement } from "react";
+import { useIntersectionObserver } from "usehooks-ts";
+
+export default function IntersectionObserverWrapper({
+  intersectionFunction,
```
Comment (FranckG28) [inline components/home/intersection-observer-wrapper.tsx:5]: Maybe `callback` is a simpler name ?

File: components/home/intersection-observer-wrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  });
+  const isVisible = !!entry?.isIntersecting;
+  useEffect(() => {
+    if (isVisible && refedChildren.props?.id) {
+      intersectionFunction(refedChildren.props.id);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isVisible]);
+
+  return <>{refedChildren}</>;
```
Comment (FranckG28) [inline components/home/intersection-observer-wrapper.tsx:29]:
```suggestion
  return refedChildren;
```
doesn't work ?

**Review by FranckG28 — COMMENTED** [review]
very small comments 😶‍🌫️

**Review by FranckG28 — APPROVED** [review]
like

**Pull request merged**